### PR TITLE
Added timestamp to EmscriptenTouchEvent

### PR DESCRIPTION
**Pull request by mad-rain**
Hi!

This PR adds timestamp field to touch event, timestamp is very useful for measuring velocity during dragging with deceleration effect.

### CODE REVIEW

File: src/library_html5.js [hunk 1]
```diff
@@ -2080,6 +2080,7 @@ var LibraryJSEvents = {
       var touchEvent = JSEvents.touchEvent;
 #endif
       var idx = touchEvent>>2; // Pre-shift the ptr to index to HEAP32 to save code size
+      {{{ makeSetValue('touchEvent', C_STRUCTS.EmscriptenTouchEvent.timestamp, 'e.timeStamp', 'double') }}};
```
Comment (kripken) [inline src/library_html5.js:2083]:
```suggestion
      {{{ makeSetValue('touchEvent', C_STRUCTS.EmscriptenTouchEvent.timestamp, 'e.timeStamp', 'double') }}};
      var idx = touchEvent>>2; // Pre-shift the ptr to index to HEAP32 to save code size
```
Reversing the order keeps `idx` next to its uses.

**Review by kripken — COMMENTED** [review]
Thanks, I think this makes sense given we already have it for EmscriptenGamepadEvent.

Please also update the docs in `site/source/docs/api_reference/html5.h.rst` (can copy the line for EmscriptenGamepadEvent.

**Comment by mad-rain** [comment]
Hi Alon!

Thanks for the feedback. I have added corresponding line to html5.h.rst.

**Comment by juj** [comment]
This would go against the direction that was set earlier in https://github.com/emscripten-core/emscripten/pull/9971 and https://github.com/emscripten-core/emscripten/pull/9987 .

If you need a timestamp, you can call `emscripten_get_now()` to get one.

**Comment by mad-rain** [comment]
1) According to this document timeStamp in Event is already have reduced time precision:
https://developer.mozilla.org/en-US/docs/Web/API/Event/timeStamp

2) Also, Emscripten is not limiting you to call anything in JS and you will still be able to receive timestamp for events in some nonstandard way.

3) emscripten_get_now gets current time, not time of an touch event. The differences will be noticeable on low-end devices, when an app is not able to run at 60 FPS.

**Comment by juj** [comment]
> 3\. emscripten_get_now gets current time, not time of an touch event.

In the non-threads case, the event handler synchronously dispatches to the callback that processes the event, so these two times will be the same, or only differ by a fraction of a nanosecond on a modern PC.

If a solution is needed for the threaded dispatch case, I think it would be better to solve this via a general solution that covers all types of events, not only touch events specifically.

**Comment by mad-rain** [comment]
It's clear for me that Emscripten dispatches a touch event to a handler immediatelly. I don't know how exactly browsers handles events internally, but usually native events pushes to an system event queue and then an app pull them from that queue. So system pushes an event with timestamp when the event has happened. The app pulls events from the queue when the app is not busy. If the app runs slowly, then difference between timestamp from the event and current time won't be zero. The more slower the app runs, the larger difference we will get.

I mean that Event.timestamp is not the same as performance.now() and shouldn't be replaced with it.

**Comment by juj** [comment]
Oh apologies, I had glanced the PR and thought the code was calling `performance.now()` to populate the time stamp.

The last time I looked there was no `timeStamp` field on TouchEvent spec - but now taking another look I realize that the `.timeStamp` field comes from the base `Event` specification at https://developer.mozilla.org/en-US/docs/Web/API/Event/timeStamp .

In that light, I think it would be good to generally marshal the passed timestamp through from all HTML5 events that are generated, not just from touch. It could indeed differ from `performance.now()`. I presume browsers do actually fill the time stamp for all input events they pass(?)

**Comment by kripken** [comment]
Adding it to all events sounds good to me. The browser provides the info, so we may as well use it.

I had assumed since this only existed in html5.h in `EmscriptenGamepadEvent` that it was particularly useful there because of how that event is used. And seems like Touch events might be similar (also lots of quick events tracking motion). But I agree it could be used in others too (mouse events?), so let's try to do them all.

**Comment by mad-rain** [comment]
I have added timestamp to EmscriptenMouseEvent & EmscriptenKeyboadEvent. I think that there are not practical reasons to add timestamps to other kind of events.

**Review by kripken — APPROVED** [review]
Thanks, looks good!

@juj did you want to take a look?

**Review by juj — APPROVED** [review]
Looks great!

**Pull request merged**